Change a commit_if_ok call to probe

JulianKnodt opened the pull request:
Removes an over-eager `commit_if_ok` which makes inference worse.

I'm not entirely sure whether it's ok to remove the check that types are the same, because casting seems to cause equality checks with incorrect types?

Fixes #105037

r? @BoxyUwU

JulianKnodt committed:
Change commit_if_ok to probe

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3618733930/jobs/6098941075) [(plain)](https://github.com/rust-lang/rust/commit/e91e7311cb5676069913035fc2d73a4998815f8b/checks/9884889094/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/compiler/rustc_trait_selection/src/traits/const_evaluatable.rs at line 156:
                 ControlFlow::BREAK
             } else if let ty::ConstKind::Expr(e) = c.kind() {
                 if let ty::Expr::Cast(_cast, ct, ty) = e {
-                    if ty == self.ct.ty()
-                            .infcx
-                            .infcx
-                            .can_eq(self.param_env, ct, self.ct)
-                            .is_ok()
+                    if ty == self.ct.ty() && self.infcx.can_eq(self.param_env, ct, self.ct).is_ok()
                         ControlFlow::BREAK
                     } else {
                     } else {
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_trait_selection/src/traits/engine.rs" "/checkout/compiler/rustc_trait_selection/src/traits/util.rs" "/checkout/compiler/rustc_trait_selection/src/traits/specialize/mod.rs" "/checkout/compiler/rustc_trait_selection/src/traits/const_evaluatable.rs" "/checkout/compiler/rustc_trait_selection/src/traits/specialize/specialization_graph.rs" "/checkout/compiler/rustc_trait_selection/src/traits/mod.rs" "/checkout/compiler/rustc_trait_selection/src/traits/wf.rs" "/checkout/compiler/rustc_errors/src/json/tests.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

BoxyUwU commented:
replacing this with a `probe` seems right, we dont want to constrain the infer vars in `self.ct` unless we're sure there are no other things that we could have unified with to consider it evaluatable.  It's still not quite right though, we do have to actually constrain the infer vars that `eq` constraints at some point in order to ensure that the const _is_ actually evaluatable. 

What's needed is to have the visitor visit all the consts regardless of it succeeds in unifying with something, and if it succeed in unifying with something (in the probe) track that in `Visitor`. Once the visitor is finished you want to be able to see of all the consts that we were able to unify with, and if there was only one, redo the `eq` and `select_all_or_error()` outside of a probe.

JulianKnodt committed:
Change based on comments

Instead of just switching to a probe, check for different matches, and see how many there are.
If one, unify it, otherwise return true and let it be unified later.

JulianKnodt commented:
I updated to check for only a single match and commit the change if there is a single one, but somehow this results in an error for the original test case. Is this an issue with the way I implemented it?

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3684737651/jobs/6234831440) [(plain)](https://github.com/rust-lang/rust/commit/0bec1bb9abd8d7f3f4f95c8619701a0806622563/checks/10060803702/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [ui] src/test/ui/const-generics/issues/issue-105037.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/const-generics/issues/issue-105037.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "-O" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/issues/issue-105037/a" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/issues/issue-105037/auxiliary"
stdout: none
--- stderr -------------------------------
  --> /checkout/src/test/ui/const-generics/issues/issue-105037.rs:32:19
   |
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
LL |     process_table(Table2::<D>);
   |     ------------- ^^^^^^^^^^^ expected struct `Table1`, found struct `Table2`
   |     arguments to this function are incorrect
   |
   |
   = note: expected struct `Table1<D>`
              found struct `Table2<D>`
  --> /checkout/src/test/ui/const-generics/issues/issue-105037.rs:20:4
   |
   |
LL | fn process_table<T: Table<D>, const D: usize>(_table: T)

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
```

</details>

BoxyUwU commented:
the current code only works for a single `ConstEvaluatable` predicate in the paramenv, i.e. if you had `where [(); N +1 + 1]:` and were trying to prove `ConstEvalutable(N + _)` it'd avoid unifying with `1 + 1` or `1` because either would work. 
The test case you're working with has two separate things in param env: `ConstEvaluatable(Table1<D>)` and `ConstEvaluatable(Table2<D>)` so you'll also have to update the for loop to make you a single `Option<Result<Const<'tcx>>, ()>` and do the `ocx.eq` stuff after the for loop instead of inside

BoxyUwU commented on compiler/rustc_trait_selection/src/traits/const_evaluatable.rs:
```diff
@@ -200,22 +199,37 @@ fn satisfied_from_param_env<'tcx>(
+                .is_ok();
+            assert!(is_ok);
+        }
+        return true;
```
You shouldn't return true if you havent rerun the `ocx.eq` and `ocx.select_all_or_error` this will lead to us considering stuff to be evaluatable that actually isnt:
```rust
#![feature(generic_const_exprs)]
#![allow(incomplete_features)]

fn foo<const N: usize, const M: usize>() -> [(); N + 2]
where
    [(); N + 1]:,
    [(); M + 1]:,
{
    bar()
}

fn bar<const N: usize>() -> [(); N]
where
    [(); N + 1]:,
{
    [(); N]
}

fn main() {}
```
this ICEs with your PR and will be fixed by doing `if let Some(Ok(c)) = single_match {` instead, so that you only return true if there was a single match and you have constrained all the infer vars

JulianKnodt replied:
While this change does fix this specific case, it does make other cases not evaluatable when they should be.
I'm not sure there's an easy way to fix it without changing more in other places? I tried to return `MentionsInfer` but that didn't seem to work

JulianKnodt committed:
Set !const_evaluatable if ambig. and not inferred

This prevents an ICE due to a value not actually being evaluatable later.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3737309859/jobs/6342395366) [(plain)](https://github.com/rust-lang/rust/commit/74505798acfd827dc3a1aaf0c4ded8e83e3e9b27/checks/10198146583/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
...i...........ii....................................................................... 1936/14029
........................................................................................ 2024/14029
............................................................................i........... 2112/14029
........................................................................................ 2200/14029
.........................................................FF.F.........F.F............... 2288/14029
.........F.........F.....FF.F.........F................................................. 2376/14029
......................................................................................F. 2464/14029
........................................................................................ 2640/14029
........................................................................................ 2728/14029
........................................................................................ 2816/14029
........................................................................................ 2904/14029
---
failures:

---- [ui] src/test/ui/const-generics/generic_const_exprs/cross_crate.rs stdout ----

error: auxiliary build of "/checkout/src/test/ui/const-generics/generic_const_exprs/auxiliary/const_evaluatable_lib.rs" failed to compile: 
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/const-generics/generic_const_exprs/auxiliary/const_evaluatable_lib.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/generic_const_exprs/cross_crate/auxiliary" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--crate-type" "dylib" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/generic_const_exprs/cross_crate/auxiliary"
stdout: none
--- stderr -------------------------------
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/auxiliary/const_evaluatable_lib.rs:6:10
   |
   |
LL |     [u8; std::mem::size_of::<T>() - 1]: Sized,
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); std::mem::size_of::<T>() - 1]:`
note: required by a bound in `test1`
   |
   |
LL | pub fn test1<T>() -> [u8; std::mem::size_of::<T>() - 1]
LL | where
LL | where
LL |     [u8; std::mem::size_of::<T>() - 1]: Sized,
   |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `test1`
error: unconstrained generic constant
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/auxiliary/const_evaluatable_lib.rs:6:10
   |
   |
LL |     [u8; std::mem::size_of::<T>() - 1]: Sized,
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); std::mem::size_of::<T>() - 1]:`
note: required by a bound in `test1`
   |
   |
LL | pub fn test1<T>() -> [u8; std::mem::size_of::<T>() - 1]
   |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `test1`
error: aborting due to 2 previous errors
------------------------------------------



---- [ui] src/test/ui/const-generics/generic_const_exprs/cross_crate_predicate.rs stdout ----

error: auxiliary build of "/checkout/src/test/ui/const-generics/generic_const_exprs/auxiliary/const_evaluatable_lib.rs" failed to compile: 
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/const-generics/generic_const_exprs/auxiliary/const_evaluatable_lib.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/generic_const_exprs/cross_crate_predicate/auxiliary" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--crate-type" "dylib" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/generic_const_exprs/cross_crate_predicate/auxiliary"
stdout: none
--- stderr -------------------------------
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/auxiliary/const_evaluatable_lib.rs:6:10
   |
   |
LL |     [u8; std::mem::size_of::<T>() - 1]: Sized,
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); std::mem::size_of::<T>() - 1]:`
note: required by a bound in `test1`
   |
   |
LL | pub fn test1<T>() -> [u8; std::mem::size_of::<T>() - 1]
LL | where
LL | where
LL |     [u8; std::mem::size_of::<T>() - 1]: Sized,
   |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `test1`
error: unconstrained generic constant
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/auxiliary/const_evaluatable_lib.rs:6:10
   |
   |
LL |     [u8; std::mem::size_of::<T>() - 1]: Sized,
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); std::mem::size_of::<T>() - 1]:`
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
note: required by a bound in `test1`
   |
   |
LL | pub fn test1<T>() -> [u8; std::mem::size_of::<T>() - 1]
   |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `test1`
error: aborting due to 2 previous errors
------------------------------------------



---- [ui] src/test/ui/const-generics/generic_const_exprs/const_eval_resolve_canonical.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/const-generics/generic_const_exprs/const_eval_resolve_canonical.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/generic_const_exprs/const_eval_resolve_canonical" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/generic_const_exprs/const_eval_resolve_canonical/auxiliary"
stdout: none
--- stderr -------------------------------
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/const_eval_resolve_canonical.rs:20:13
   |
   |
LL |     (): Foo<{ N + 1 }>,
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); { N + 1 }]:`
note: required by a bound in `foo`
   |
   |
LL | fn foo<T, const N: usize>(_: T) -> <() as Foo<{ N + 1 }>>::Assoc
LL | where
LL | where
LL |     (): Foo<{ N + 1 }>,
   |             ^^^^^^^^^ required by this bound in `foo`
error: unconstrained generic constant
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/const_eval_resolve_canonical.rs:20:13
   |
   |
LL |     (): Foo<{ N + 1 }>,
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); { N + 1 }]:`
note: required by a bound in `foo`
   |
   |
LL | fn foo<T, const N: usize>(_: T) -> <() as Foo<{ N + 1 }>>::Assoc
   |                                               ^^^^^^^^^ required by this bound in `foo`
error: unconstrained generic constant
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/const_eval_resolve_canonical.rs:20:9
   |
   |
LL |     (): Foo<{ N + 1 }>,
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); { N + 1 }]:`
error: unconstrained generic constant
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/const_eval_resolve_canonical.rs:18:36
   |
   |
LL | fn foo<T, const N: usize>(_: T) -> <() as Foo<{ N + 1 }>>::Assoc
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); { N + 1 }]:`
error: aborting due to 4 previous errors
------------------------------------------



---- [ui] src/test/ui/const-generics/generic_const_exprs/fn_call.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/const-generics/generic_const_exprs/fn_call.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "-O" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/generic_const_exprs/fn_call/a" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/generic_const_exprs/fn_call/auxiliary"
stdout: none
--- stderr -------------------------------
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/fn_call.rs:15:10
   |
   |
LL |     [u8; std::mem::size_of::<T>()]: Sized,
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); std::mem::size_of::<T>()]:`
note: required by a bound in `test_simple`
   |
   |
LL | fn test_simple<T>() -> [u8; std::mem::size_of::<T>()]
LL | where
LL | where
LL |     [u8; std::mem::size_of::<T>()]: Sized,
   |          ^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `test_simple`
error: unconstrained generic constant
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/fn_call.rs:15:10
   |
   |
LL |     [u8; std::mem::size_of::<T>()]: Sized,
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); std::mem::size_of::<T>()]:`
note: required by a bound in `test_simple`
   |
   |
LL | fn test_simple<T>() -> [u8; std::mem::size_of::<T>()]
   |                             ^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `test_simple`
error: unconstrained generic constant
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/fn_call.rs:22:10
   |
   |
LL |     [u8; test_me::<T>(N, N + 1) + N]: Sized,
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); test_me::<T>(N, N + 1) + N]:`
note: required by a bound in `test_with_args`
   |
   |
LL | fn test_with_args<T, const N: usize>() -> [u8; test_me::<T>(N, N + 1) + N]
LL | where
LL | where
LL |     [u8; test_me::<T>(N, N + 1) + N]: Sized,
   |          ^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `test_with_args`
error: unconstrained generic constant
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/fn_call.rs:22:10
   |
   |
LL |     [u8; test_me::<T>(N, N + 1) + N]: Sized,
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); test_me::<T>(N, N + 1) + N]:`
note: required by a bound in `test_with_args`
   |
   |
LL | fn test_with_args<T, const N: usize>() -> [u8; test_me::<T>(N, N + 1) + N]
   |                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `test_with_args`
error: aborting due to 4 previous errors
------------------------------------------



---- [ui] src/test/ui/const-generics/generic_const_exprs/infer-too-generic.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/const-generics/generic_const_exprs/infer-too-generic.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "-O" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/generic_const_exprs/infer-too-generic/a" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/generic_const_exprs/infer-too-generic/auxiliary"
stdout: none
--- stderr -------------------------------
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/infer-too-generic.rs:9:9
   |
   |
LL |     [T; N - 1]: Sized,
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); N - 1]:`
note: required by a bound in `split_first`
   |
   |
LL | fn split_first<T, const N: usize>(arr: [T; N]) -> (T, [T; N - 1])
LL | where
LL | where
LL |     [T; N - 1]: Sized,
   |         ^^^^^ required by this bound in `split_first`
error: unconstrained generic constant
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/infer-too-generic.rs:9:9
   |
   |
LL |     [T; N - 1]: Sized,
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); N - 1]:`
note: required by a bound in `split_first`
   |
   |
LL | fn split_first<T, const N: usize>(arr: [T; N]) -> (T, [T; N - 1])
   |                                                           ^^^^^ required by this bound in `split_first`
error: aborting due to 2 previous errors
------------------------------------------



---- [ui] src/test/ui/const-generics/generic_const_exprs/issue-84408.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/const-generics/generic_const_exprs/issue-84408.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/generic_const_exprs/issue-84408" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/generic_const_exprs/issue-84408/auxiliary"
stdout: none
--- stderr -------------------------------
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/issue-84408.rs:28:5
   |
   |
LL |     [(); A + B]: ,
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); { A + B }]:`
note: required by a bound in `<Bar<A, B> as Melon<{ A + B }>>`
   |
   |
LL | impl<const A: usize, const B: usize> Melon<{ A + B }> for Bar<A, B>
   |                                            ^^^^^^^^^ required by this bound in `<Bar<A, B> as Melon<{ A + B }>>`
error: unconstrained generic constant
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/issue-84408.rs:28:5
   |
   |
LL |     [(); A + B]: ,
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); A + B]:`
note: required by a bound in `<Bar<A, B> as Melon<{ A + B }>>`
   |
   |
LL |     [(); A + B]: ,
   |          ^^^^^ required by this bound in `<Bar<A, B> as Melon<{ A + B }>>`
error: unconstrained generic constant
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/issue-84408.rs:26:59
   |
   |
LL | impl<const A: usize, const B: usize> Melon<{ A + B }> for Bar<A, B>
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); A + B]:`
note: required by a bound in `Bar`
   |
   |
LL | struct Bar<const A: usize, const B: usize>([i32; A + B])
LL | where
LL | where
LL |     [(); A + B]: ;
   |          ^^^^^ required by this bound in `Bar`
error: unconstrained generic constant
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/issue-84408.rs:26:38
   |
   |
LL | impl<const A: usize, const B: usize> Melon<{ A + B }> for Bar<A, B>
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); { A + B }]:`
error: unconstrained generic constant
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/issue-84408.rs:28:5
   |
   |
LL |     [(); A + B]: ,
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); { A + B }]:`
note: required by a bound in `<Bar<A, B> as Melon<{ A + B }>>`
   |
   |
LL | impl<const A: usize, const B: usize> Melon<{ A + B }> for Bar<A, B>
   |                                            ^^^^^^^^^ required by this bound in `<Bar<A, B> as Melon<{ A + B }>>`
error: unconstrained generic constant
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/issue-84408.rs:28:5
   |
   |
LL |     [(); A + B]: ,
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); A + B]:`
note: required by a bound in `<Bar<A, B> as Melon<{ A + B }>>`
   |
   |
LL |     [(); A + B]: ,
   |          ^^^^^ required by this bound in `<Bar<A, B> as Melon<{ A + B }>>`
error: unconstrained generic constant
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/issue-84408.rs:30:23
   |
   |
LL |     fn new(arr: [i32; A + B]) -> Self {
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); A + B]:`
note: required by a bound in `<Bar<A, B> as Melon<{ A + B }>>::new`
   |
   |
LL |     fn new(arr: [i32; A + B]) -> Self {
   |                       ^^^^^ required by this bound in `<Bar<A, B> as Melon<{ A + B }>>::new`
error: unconstrained generic constant
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/issue-84408.rs:30:34
   |
   |
LL |     fn new(arr: [i32; A + B]) -> Self {
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); A + B]:`
note: required by a bound in `Bar`
   |
   |
LL | struct Bar<const A: usize, const B: usize>([i32; A + B])
LL | where
LL | where
LL |     [(); A + B]: ;
   |          ^^^^^ required by this bound in `Bar`
error: unconstrained generic constant
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/issue-84408.rs:28:5
   |
   |
LL |     [(); A + B]: ,
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); { A + B }]:`
note: required by a bound in `<Bar<A, B> as Melon<{ A + B }>>`
   |
   |
LL | impl<const A: usize, const B: usize> Melon<{ A + B }> for Bar<A, B>
   |                                            ^^^^^^^^^ required by this bound in `<Bar<A, B> as Melon<{ A + B }>>`
error: unconstrained generic constant
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/issue-84408.rs:28:5
   |
   |
LL |     [(); A + B]: ,
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); A + B]:`
note: required by a bound in `<Bar<A, B> as Melon<{ A + B }>>`
   |
   |
LL |     [(); A + B]: ,
   |          ^^^^^ required by this bound in `<Bar<A, B> as Melon<{ A + B }>>`
error: unconstrained generic constant
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/issue-84408.rs:33:24
   |
   |
LL |     fn change<T: Melon<{ A + B }>>(self) -> T {
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); { A + B }]:`
note: required by a bound in `<Bar<A, B> as Melon<{ A + B }>>::change`
   |
   |
LL |     fn change<T: Melon<{ A + B }>>(self) -> T {
   |                        ^^^^^^^^^ required by this bound in `<Bar<A, B> as Melon<{ A + B }>>::change`
error: unconstrained generic constant
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/issue-84408.rs:33:36
   |
   |
LL |     fn change<T: Melon<{ A + B }>>(self) -> T {
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); A + B]:`
note: required by a bound in `Bar`
   |
   |
LL | struct Bar<const A: usize, const B: usize>([i32; A + B])
LL | where
LL | where
LL |     [(); A + B]: ;
   |          ^^^^^ required by this bound in `Bar`
error: unconstrained generic constant
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/issue-84408.rs:33:18
   |
   |
LL |     fn change<T: Melon<{ A + B }>>(self) -> T {
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); { A + B }]:`
error: aborting due to 13 previous errors
------------------------------------------



---- [ui] src/test/ui/const-generics/generic_const_exprs/less_than.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/const-generics/generic_const_exprs/less_than.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "-O" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/generic_const_exprs/less_than/a" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/generic_const_exprs/less_than/auxiliary"
stdout: none
--- stderr -------------------------------
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/less_than.rs:7:56
   |
   |
LL | fn test<const N: usize>() -> Foo<{ N > 10 }> where Foo<{ N > 10 }>: Sized {
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); { N > 10 }]:`
note: required by a bound in `test`
   |
   |
LL | fn test<const N: usize>() -> Foo<{ N > 10 }> where Foo<{ N > 10 }>: Sized {
   |                                                        ^^^^^^^^^^ required by this bound in `test`
error: unconstrained generic constant
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/less_than.rs:7:56
   |
   |
LL | fn test<const N: usize>() -> Foo<{ N > 10 }> where Foo<{ N > 10 }>: Sized {
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); { N > 10 }]:`
note: required by a bound in `test`
   |
   |
LL | fn test<const N: usize>() -> Foo<{ N > 10 }> where Foo<{ N > 10 }>: Sized {
   |                                  ^^^^^^^^^^ required by this bound in `test`
error: unconstrained generic constant
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/less_than.rs:7:30
   |
   |
LL | fn test<const N: usize>() -> Foo<{ N > 10 }> where Foo<{ N > 10 }>: Sized {
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); { N > 10 }]:`
error: unconstrained generic constant
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/less_than.rs:7:69
   |
   |
LL | fn test<const N: usize>() -> Foo<{ N > 10 }> where Foo<{ N > 10 }>: Sized {
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); { N > 10 }]:`
error: aborting due to 4 previous errors
------------------------------------------



---- [ui] src/test/ui/const-generics/generic_const_exprs/nested_uneval_unification-1.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/const-generics/generic_const_exprs/nested_uneval_unification-1.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "-O" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/generic_const_exprs/nested_uneval_unification-1/a" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/generic_const_exprs/nested_uneval_unification-1/auxiliary"
stdout: none
--- stderr -------------------------------
  --> /checkout/src/test/ui/const-generics/generic_const_exprs/nested_uneval_unification-1.rs:17:5
   |
   |
LL |     [(); { M * 2 } + 1]: ,
   |
   |
   = help: try adding a `where` bound using this expression: `where [(); { M * 2 }]:`
note: required by a bound in `substs2`
   |
   |
LL | fn substs2<const M: usize>() -> Substs1<{ M * 2 }>
   |                                         ^^^^^^^^^ required by this bound in `substs2`
error: unconstrained generic constant
```

</details>

BoxyUwU commented:
I think all the CI failures are caused by cases where we have duplicate candidates, i.e. two `N + 1` in the param env, you probably want to avoid setting `single_match` to `Err` instead of `Ok` if the candidate that matched is `==` to the existing `Ok`

JulianKnodt committed:
Check for duplicates

BoxyUwU commented:
Can you add the following as tests:

---
```rust
#![feature(generic_const_exprs)]
#![allow(incomplete_features)]

fn foo<const N: usize>()
where
    [(); N + 1]:,
    [(); N + 1]:,
{
    bar::<N>();
}

fn bar<const N: usize>()
where
    [(); N + 1]:,
{
}

fn main() {}
```
---
```rust
#![feature(generic_const_exprs)]
#![allow(incomplete_features)]

const fn both(_: usize, b: usize) -> usize {
    b
}

fn foo<const N: usize>()
where
    [(); both(N + 1, N + 1)]:,
{
    bar::<N>();
}

fn bar<const N: usize>()
where
    [(); N + 1]:,
{
}

fn main() {}
```
---
```rust
#![feature(generic_const_exprs)]
#![allow(incomplete_features)]

const fn both(_: usize, b: usize) -> usize {
    b
}

fn foo<const N: usize, const M: usize>() -> [(); N + 2]
where
    [(); both(N + 1, M + 1)]:,
{
    bar()
}

fn bar<const N: usize>() -> [(); N]
where
    [(); N + 1]:,
{
    [(); N]
}

fn main() {}
```
---

BoxyUwU commented on compiler/rustc_trait_selection/src/traits/const_evaluatable.rs:
```diff
@@ -200,22 +202,44 @@ fn satisfied_from_param_env<'tcx>(
+                let mut v = Visitor { ct, infcx, param_env, single_match: None };
+                let _ = b_ct.visit_with(&mut v);
+
+                if let Some(inner) = v.single_match {
+                    single_match = if let Ok(inner) = inner {
+                        match single_match {
+                            None => Some(Ok(inner)),
+                            Some(Ok(prev)) if prev == inner => Some(Ok(prev)),
+                            Some(_) => Some(Err(())),
+                        }
+                    } else {
+                        Some(Err(()))
+                    };
```
```suggestion
                single_match = match v.single_match {
                    None => single_match,
                    Some(Ok(c)) => match single_match {
                        None => Some(Ok(c)),
                        Some(Ok(prev)) if prev == c => Some(Ok(prev)),
                        Some(_) => Some(Err(())),
                    },
                    Some(Err(())) => Some(Err(())),
                };
```

BoxyUwU commented on compiler/rustc_trait_selection/src/traits/const_evaluatable.rs:
```diff
@@ -200,22 +202,44 @@ fn satisfied_from_param_env<'tcx>(
     }
 
+    if let Some(Ok(c)) = single_match {
+        let is_ok = infcx
+            .commit_if_ok(|_| {
+                let ocx = ObligationCtxt::new_in_snapshot(infcx);
+                assert!(ocx.eq(&ObligationCause::dummy(), param_env, c.ty(), ct.ty()).is_ok());
+                assert!(ocx.eq(&ObligationCause::dummy(), param_env, c, ct).is_ok());
+                if ocx.select_all_or_error().is_empty() { Ok(()) } else { Err(()) }
+            })
+            .is_ok();
+        assert!(is_ok);
```
```suggestion
        let ocx = ObligationCtxt::new(infcx);
        assert!(ocx.eq(&ObligationCause::dummy(), param_env, c.ty(), ct.ty()).is_ok());
        assert!(ocx.eq(&ObligationCause::dummy(), param_env, c, ct).is_ok());
        assert!(ocx.select_all_or_error().is_empty());
```

BoxyUwU commented on compiler/rustc_trait_selection/src/traits/const_evaluatable.rs:
```diff
@@ -168,23 +168,25 @@ fn satisfied_from_param_env<'tcx>(
+                    Some(Ok(o)) if o == c => Some(Ok(c)),
+                    Some(_) => Some(Err(())),
+                };
+                ControlFlow::CONTINUE
             } else if let ty::ConstKind::Expr(e) = c.kind() {
```
```suggestion
            }
            
            if let ty::ConstKind::Expr(e) = c.kind() {
```

BoxyUwU replied:
And adding this as a test:
```rust
#![feature(generic_const_exprs)]
#![allow(incomplete_features)]

fn foo<const N: usize>()
where
    [(); N + 1 + 1]:,
{
    bar();
}

fn bar<const N: usize>()
where
    [(); N + 1]:,
{
}

fn main() {}
```

JulianKnodt committed:
Clean up

Simplify match statement

Add multiple tests
- 1 test for checking `N + 1 + 1` does not unify with `N+1`
- 2 tests for checking that a function that uses two parameters only returns the parameter that
  is actually used.
- Check exact repeat predicates

BoxyUwU commented:
will r+ this once its rebased

BoxyUwU commented:
@bors r+ rollup

bors commented:
:pushpin: Commit 21c5ffe008cce39bcd676ed197f691adbfbf7a2f has been approved by `BoxyUwU`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=BoxyUwU 21c5ffe008cce39bcd676ed197f691adbfbf7a2f -->
<!-- homu: {"type":"Approved","sha":"21c5ffe008cce39bcd676ed197f691adbfbf7a2f","approver":"BoxyUwU","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
